### ISSUE TITLE
Have Flathub as the sole distribution mechanism for MMEX on Linux

### ISSUE THREAD

**Issue by whalley** [moneymanagerex/moneymanagerex#6246]
Are we happy have Flathub as the sole distribution mechanism for MMEX on Linux, and remove the other Linux builds we make via CircleCI. I'm concious that the CircleCI builds are still on wx3.1.7 and are falling behind Linux OS release versions with no active maintenance.

Look to include Flatpak build as part of CI and drop other LInux support.

_Originally posted by @whalley in https://github.com/moneymanagerex/moneymanagerex/issues/5989#issuecomment-1761316530_

**Comment by whalley** [comment]
Added  `--privileged` to the build and it builds wxWidgets but then fails.. 

```
#!/bin/bash -eo pipefail
docker run -it --rm --privileged -w /moneymanagerex/build -v $HOME/.ccache:/root/.ccache -v $CIRCLE_WORKING_DIRECTORY:/moneymanagerex bilelmoussaoui/flatpak-github-actions:gnome-45 bash -c "rm -rf * && /usr/bin/xvfb-run --auto-servernum flatpak-builder --repo=repo --disable-rofiles-fuse --force-clean --default-branch=master --arch=x86_64 _build ../org.moneymanagerex.MMEX.yml && /usr/bin/xvfb-run --auto-servernum flatpak build-bundle --arch=x86_64 ../mmex.flatpak org.moneymanagerex.MMEX master" && \
cp $CIRCLE_WORKING_DIRECTORY/build/mmex*.flatpak /tmp/artifacts
```

https://app.circleci.com/pipelines/github/whalley/circleci-test/141/workflows/1a4c957f-38f8-4713-a5cf-29587f5780cf/jobs/171
```
.
.
.
.
------------------------------------------------------
 
 The installation of wxWidgets is finished.  On certain
 platforms (e.g. Linux) you'll now have to run ldconfig
 if you installed a shared library and also modify the
 LD_LIBRARY_PATH (or equivalent) environment variable.
 
 wxWidgets comes with no guarantees and doesn't claim
 to be suitable for any purpose.
 
 Read the wxWindows Licence on licencing conditions.
 
 ------------------------------------------------------
 
compressing debuginfo in: /moneymanagerex/build/_build/files/lib/wx/3.2/web-extensions/webkit2_extu-3.2.so
processing: /moneymanagerex/build/_build/files/lib/wx/3.2/web-extensions/webkit2_extu-3.2.so
[29] .debug_aranges compressed -> .zdebug_aranges (192 => 79 41.15%)
[30] .debug_info compressed -> .zdebug_info (17178 => 9885 57.54%)
[31] .debug_abbrev compressed -> .zdebug_abbrev (1636 => 732 44.74%)
[32] .debug_line compressed -> .zdebug_line (2527 => 1255 49.66%)
[33] .debug_str compressed -> .zdebug_str (20045 => 7158 35.71%)
[34] .debug_line_str compressed -> .zdebug_line_str (1551 => 627 40.43%)
[35] .debug_loclists compressed -> .zdebug_loclists (3855 => 1571 40.75%)
[36] .debug_rnglists compressed -> .zdebug_rnglists (301 => 207 68.77%)
[39] Updating section string table
stripping /moneymanagerex/build/_build/files/lib/wx/3.2/web-extensions/webkit2_extu-3.2.so to /moneymanagerex/build/_build/files/lib/debug/lib/wx/3.2/web-extensions/webkit2_extu-3.2.so.debug
compressing debuginfo in: /moneymanagerex/build/_build/files/bin/wxrc-3.2
processing: /moneymanagerex/build/_build/files/bin/wxrc-3.2
[31] .debug_aranges compressed -> .zdebug_aranges (38112 => 6252 16.40%)
[32] .debug_info compressed -> .zdebug_info (17737908 => 7864178 44.34%)
[33] .debug_abbrev compressed -> .zdebug_abbrev (359619 => 78295 21.77%)
[34] .debug_line compressed -> .zdebug_line (2633298 => 662204 25.15%)
[35] .debug_str compressed -> .zdebug_str (768628 => 152528 19.84%)
[36] .debug_line_str compressed -> .zdebug_line_str (5208 => 1765 33.89%)
[37] .debug_loclists compressed -> .zdebug_loclists (5159606 => 1500635 29.08%)
[38] .debug_rnglists compressed -> .zdebug_rnglists (1053715 => 370371 35.15%)
[41] Updating section string table
stripping /moneymanagerex/build/_build/files/bin/wxrc-3.2 to /moneymanagerex/build/_build/files/lib/debug/bin/wxrc-3.2.debug
Committing stage build-wxWidgets to cache
========================================================================
Building module moneymanagerex in /moneymanagerex/build/.flatpak-builder/build/moneymanagerex-1
========================================================================
Error: module moneymanagerex: Error opening file /proc/sys/net/ipv4/route/flush: Permission denied

Exited with code exit status 1
```

**Comment by joshua-stone** [comment]
@whalley Looks like the build made a bit of progress.

Can you modify [this line at the bottom](https://github.com/whalley/circleci-test/blob/a32b35c0105c6a88db44e6a68d6e2191b8763992/org.moneymanagerex.MMEX.yml#L52) from `path: ..` to `path: .`, and add `--verbose` to flatpak-builder so we can see more of what it's doing?

**Comment by whalley** [comment]
Seems better....
https://app.circleci.com/pipelines/github/whalley/circleci-test/142/workflows/53cc13c9-6559-4ff4-9908-b3fb8b7e4f82/jobs/172

**Comment by joshua-stone** [comment]
Much better!

It looks like `flatpak build-bundle` is missing an argument [as shown here](https://github.com/MarvellousSoft/MarvInc/blob/27845edbfb421761112a01a697016c62a39f89dc/deploy.sh#L268). Can you try using this?

```
flatpak build-bundle --arch=x86_64 repo ../mmex.flatpak org.moneymanagerex.MMEX master
```

**Comment by whalley** [comment]
https://app.circleci.com/pipelines/github/whalley/circleci-test/143/workflows/88dd96a2-55dc-4c87-85a7-e56e642f0ca7/jobs/173

Just fails on the final artifact copy now. Do you know where the `mmex*.flatpak` can be found. It's not in that build directory.

**Comment by joshua-stone** [comment]
So close now! You may have to try searching for it within the container, such adding this command right after `flatpak build-bundle`:

```
find -name 'mmex*.flatpak'
```

**Comment by joshua-stone** [comment]
Looks like it installs and runs:

```
$ wget -q https://output.circle-artifacts.com/output/job/207d6cb1-2aa2-4959-9292-a43bb242f6ea/artifacts/0/tmp/artifacts/mmex.flatpak
$ flatpak install --user --assumeyes ./mmex.flatpak

org.moneymanagerex.MMEX permissions:
    ipc      network              fallback-x11      wayland      x11
    dri      file access [1]

    [1] /media, /mnt, /run/media, home


        ID                               Branch          Op          Remote              Download
 1. [✓] org.moneymanagerex.MMEX          master          i           mmex-origin         0 bytes

Installation complete.
$ flatpak info --user org.moneymanagerex.MMEX 

MMEX Money Manager - Organize finances and track cash flow

          ID: org.moneymanagerex.MMEX
         Ref: app/org.moneymanagerex.MMEX/x86_64/master
        Arch: x86_64
      Branch: master
     Version: 1.6.4
     License: GPL-2.0-only
      Origin: mmex-origin
  Collection: 
Installation: user
   Installed: 39.0 MB
     Runtime: org.gnome.Platform/x86_64/45
         Sdk: org.gnome.Sdk/x86_64/45

      Commit: 658eac2937c08a994a73b12be248c724e59b0eb3f21b630af36ad62c5e7349eb
     Subject: Export org.moneymanagerex.MMEX
        Date: 2023-10-18 15:02:41 +0000
$ flatpak run --user org.moneymanagerex.MMEX
```

You can probably test the installation process by attempting to install the .flatpak bundle inside the container:

```
$ flatpak install --bundle --noninteractive ../mmex.flatpak
$ flatpak run org.moneymanagerex.MMEX --help
```

And then have a unique version string for the build artifact:

```
$ MMEX_VERSION="$(flatpak info org.moneymanagerex.MMEX | grep Version | awk '{print $2}')"
$ cp $CIRCLE_WORKING_DIRECTORY/mmex*.flatpak /tmp/artifacts/mmex-${MMEX_VERSION}.flatpak
```

**Comment by joshua-stone** [comment]
Unfortunately, this won't succeed because it needs `flatpak install --bundle --noninteractive ../mmex.flatpak` to be included as well since `flatpak info org.moneymanagerex.MMEX` only works with installed artifacts:

https://github.com/whalley/circleci-test/commit/69ec908d3e938f786c75b89c77827ac9be7e59a8

**Comment by whalley** [comment]
Yes, just noticed. Will it work anyway as flatpack is only availabel in the container. Would we need to install flatpack in the VM or generate the correct version info in the container?

**Comment by joshua-stone** [comment]
@whalley I think it would be best to try to install it within the container as part of the build for more complete coverage. VMs shouldn't be necessary either since the container already has flatpaks installed:

```
$ podman run bilelmoussaoui/flatpak-github-actions:gnome-45 flatpak list
Mesa	org.freedesktop.Platform.GL.default	23.1.8	23.08	system
Mesa (Extra)	org.freedesktop.Platform.GL.default	23.1.8	23.08-extra	system
openh264	org.freedesktop.Platform.openh264	2.1.0	2.2.0	system
GNOME Application Platform version 45	org.gnome.Platform		45	system
GNOME Software Development Kit version 45	org.gnome.Sdk		45	system
```

Can you see if either of these lines work?

```
flatpak install --bundle --noninteractive ../mmex.flatpak
/usr/bin/xvfb-run --auto-servernum flatpak install --bundle --noninteractive ../mmex.flatpak
```

**Comment by whalley** [comment]
The first seems to work. How do you propose getting the build/commit into the filename?

**Comment by joshua-stone** [comment]
Maybe add something like this?

```
flatpak install --bundle --noninteractive ../mmex.flatpak && cp $CIRCLE_WORKING_DIRECTORY/mmex.flatpak /tmp/artifacts/mmex-$(flatpak info org.moneymanagerex.MMEX | grep Version | awk '{print $2}')-$(git rev-parse --short HEAD).flatpak
```

**Comment by whalley** [comment]
But I think the /tmp/artifacts is outside the container and not mounted as a visible filesystem. Was thinking more about just renaming as last action in container, but not sure git commit is visible there (though other builds do not include commit in name so not a deal breaker).

**Comment by joshua-stone** [comment]
You could skip using a git commit for now and make it a feature for a follow-up feature request for when we can refine the CI.

can you see if this works?

```
flatpak install --bundle --noninteractive ../mmex.flatpak && cp $CIRCLE_WORKING_DIRECTORY/mmex.flatpak /tmp/artifacts/mmex-$(flatpak info org.moneymanagerex.MMEX | grep Version | awk '{print $2}').flatpak
```

**Comment by whalley** [comment]
Guess we could rename to include the commit as part of the file copy to /tmp/artifacts outside the VM. Adding in the `$(git rev-parse --short HEAD)`

**Comment by joshua-stone** [comment]
That sounds feasible. Maybe have it as part of a bulk-renaming step so that all artifacts match a certain format?

**Comment by whalley** [comment]
Don't think rename worked... Will have a look tommorow if it's a typo on my part.
https://app.circleci.com/pipelines/github/whalley/circleci-test/151/workflows/06fd6677-93ee-4606-be92-dedb5a9c061b/jobs/181/artifacts

**Comment by joshua-stone** [comment]
TBH I think it'd be better to turn this into a script because all of these inline commands are hard to follow and not very flexible.

**Comment by joshua-stone** [comment]
Can you see what running `flatpak info org.moneymanagerex.MMEX` by itself after installing the bundle returns?

**Comment by whalley** [comment]
Actually the version number you are using is wrong, you seem to be getting it from the latest published release and not from the versioning in CMakeLists.txt.

It should be "1.7.0-Beta.1"

**Comment by joshua-stone** [comment]
> Actually the version number you are using is wrong, you seem to be getting it from the latest published release and not from the versioning in CMakeLists.txt.
> 
> It should be "1.7.0-Beta.1"

It's actually correct insofar that it's using [`org.moneymanagerex.MMEX.metainfo.xml`](https://github.com/moneymanagerex/moneymanagerex/blob/88b8131d6625348bd21c98c43032b278b663fe7f/resources/dist/linux/share/metainfo/org.moneymanagerex.MMEX.metainfo.xml#L343) as a source of truth at install time. The only builds it's using as the ones that're generated by the CI build process, not the ones published on Flathub.

Even if you generated a file like `mmex-1.7.0-Beta.1.flatpak`, it would still show up as `1.6.4` when running `flatpak info org.moneymanagerex.MMEX` because that's the latest version described in the metainfo file. That's why you have to have a process of keeping these version strings in sync so we can avoid mismatches like whenever a new release is made:

https://github.com/moneymanagerex/moneymanagerex/commits/master/resources/dist/linux/share/metainfo/org.moneymanagerex.MMEX.metainfo.xml

**Comment by whalley** [comment]
Understood, but I feel we should be generating the org.moneymanagerex.MMEX.metainfo.xml file. It feels wrong and error prone to manually maintain this.

**Comment by joshua-stone** [comment]
> Understood, but I feel we should be generating the org.moneymanagerex.MMEX.metainfo.xml file. It feels wrong and error prone to manually maintain this.

Agreed. We can make this process easier by maintaining a NEWS file that combines with the metainfo file:

https://github.com/moneymanagerex/moneymanagerex/pull/6035#issuecomment-1601905577

It should also be trivial to have the buildsystem extract the version string with something like `head -n1 NEWS` instead of hardcoding it inside [CMakeLists.txt](https://github.com/moneymanagerex/moneymanagerex/blob/8f42f5783e4b29dadbd943d8b8c901dff276123b/CMakeLists.txt#L20).

**Comment by ovari** [comment]
Will beta versions still be downloadable at https://app.circleci.com/pipelines/github/moneymanagerex/moneymanagerex ?

**Comment by whalley** [comment]
> Will beta versions still be downloadable at https://app.circleci.com/pipelines/github/moneymanagerex/moneymanagerex ?

Yes they will be, the intention is to have the flatpak available here.

**Comment by whalley** [comment]
I presume the correct version (1.7.0-Beta.1) is visible inside the About menu in MMEX when run? If so, then I suggest we go with the `mmex.flatpak` for now as the artifact name and look to resolve the full naming issue as part of resolution of https://github.com/moneymanagerex/moneymanagerex/pull/6035#issuecomment-1601905577

**Comment by joshua-stone** [comment]
Yes, the correct version is set in the test build:

![image](https://github.com/moneymanagerex/moneymanagerex/assets/8183967/b17e4f15-f4e8-4bb1-a569-9ef3ecb4a959)

**Comment by whalley** [comment]
https://github.com/moneymanagerex/moneymanagerex/pull/6249

**Comment by ovari** [comment]
> can you explain to me what is now the advantage of this solution?

MMEX will be listed the Update Manager so that users can more easily see and install the latest version.

**Comment by ovari** [comment]
1. Download `mmex.flatpak` from **Artifacts** in https://app.circleci.com/pipelines/github/moneymanagerex/moneymanagerex
2. Open Terminal in directory where the file has been downloaded
3. Type `flatpak install mmex.flatpak`
4. Answer prompts until `Installation complete.`

**Comment by renato-mmex** [comment]
@ovari  This refers to Linux and on which distri does this work, all ?
I have on a PC Linux Mint, I will test tomorrow

**Comment by ovari** [comment]
@renato-mmex

> This refers to Linux and on which distri does this work, all ?
1. https://flatpak.org/setup/
2. Click on the distro. For example, Linux Mint https://flatpak.org/setup/Linux%20Mint states:
> Flatpak support is built into Linux Mint 18.3 and newer—no setup required!

## Steps to install
1. Backup MMEX database files
2. Uninstall all versions of MMEX (deb/rpm and/or flatpak)
3. Visit https://app.circleci.com/pipelines/github/moneymanagerex/moneymanagerex
4. Click green button that states `Success`
5. Click `build-linux`
6. Click `Artifacts`
7. Under `Parallel Run` heading, click `tmp/artifacts/mmex.flatpak` to download
8. Open Terminal in directory where the `mmex.flatpak` file has been downloaded
9. Type `flatpak install mmex.flatpak`
10. Answer prompts until `Installation complete.`

**Comment by renato-mmex** [comment]
@ovari @whalley @joshua-stone 
**Installed MMEX on Linux Mint 21.1 as described above without any problems.
Thanks for the good description and good work everyone.**
